#Installing necessary libraries and versions
Make sure to restart session after installing

In [1]:
"""
!pip install -q torch
!pip install sentence-transformers==2.2.2
#!pip install -q datasets
!pip install -q faiss-cpu
!pip install -q jq
!pip install -q accelerate==0.21.0 transformers==4.31.0 tokenizers==0.13.3
!pip install -q cohere
!pip install -q tiktoken
!pip install -q openai
!pip install -q langchain
!pip install -q langchain-community langchainhub langchain-openai langsmith
!pip install -q faiss-gpu==1.7.1.post3
!pip install -q bitsandbytes==0.40.0 einops==0.6.1
!pip install -q xformers==0.0.22.post7
!pip install -q InstructorEmbedding
!pip install python-dotenv
"""

'\n!pip install -q torch\n!pip install sentence-transformers==2.2.2\n#!pip install -q datasets\n!pip install -q faiss-cpu\n!pip install -q jq\n!pip install -q accelerate==0.21.0 transformers==4.31.0 tokenizers==0.13.3\n!pip install -q cohere\n!pip install -q tiktoken\n!pip install -q openai\n!pip install -q langchain\n!pip install -q langchain-community langchainhub langchain-openai langsmith\n!pip install -q faiss-gpu==1.7.1.post3\n!pip install -q bitsandbytes==0.40.0 einops==0.6.1\n!pip install -q xformers==0.0.22.post7\n!pip install -q InstructorEmbedding\n!pip install python-dotenv\n'

#Import necessary functions from the libraries

import locale helps when sometimes encoding error occurs on installing any new library

In [2]:
#import locale
#locale.getpreferredencoding = lambda: "UTF-8"

In [3]:
from langchain_core.output_parsers import StrOutputParser
from langchain_core.runnables import RunnablePassthrough
from langchain_core.prompts import PromptTemplate
from langchain_openai import OpenAIEmbeddings
from InstructorEmbedding import INSTRUCTOR
from langchain.embeddings import HuggingFaceInstructEmbeddings, HuggingFaceEmbeddings
from langchain.document_loaders import HuggingFaceDatasetLoader
from langchain.text_splitter import RecursiveCharacterTextSplitter, CharacterTextSplitter
from langchain.vectorstores import FAISS
from transformers import AutoTokenizer, AutoModelForCausalLM, LlamaForCausalLM, AutoConfig, BitsAndBytesConfig, pipeline
from langchain_community.document_loaders import JSONLoader
from langchain import HuggingFacePipeline, smith, hub
from langchain.chains import RetrievalQA, LLMChain
from torch import cuda, bfloat16
from dotenv import load_dotenv
from pathlib import Path
from pprint import pprint
import langsmith
import warnings
import getpass
import json
import os

/usr/local/lib/python3.10/dist-packages/InstructorEmbedding/instructor.py:7: TqdmExperimentalWarning: Using `tqdm.autonotebook.tqdm` in notebook mode. Use `tqdm.tqdm` instead to force console mode (e.g. in jupyter console)
  from tqdm.autonotebook import trange



===================================BUG REPORT===================================
Welcome to bitsandbytes. For bug reports, please run

python -m bitsandbytes

 and submit this information together with your error trace to: https://github.com/TimDettmers/bitsandbytes/issues
bin /usr/local/lib/python3.10/dist-packages/bitsandbytes/libbitsandbytes_cuda122.so


/usr/local/lib/python3.10/dist-packages/bitsandbytes/cuda_setup/main.py:149: UserWarning: /usr/lib64-nvidia did not contain ['libcudart.so', 'libcudart.so.11.0', 'libcudart.so.12.0'] as expected! Searching further paths...
  warn(msg)
/usr/local/lib/python3.10/dist-packages/bitsandbytes/cuda_setup/main.py:149: UserWarning: WARNING: The following directories listed in your path were found to be non-existent: {PosixPath('/sys/fs/cgroup/memory.events /var/colab/cgroup/jupyter-children/memory.events')}
  warn(msg)
/usr/local/lib/python3.10/dist-packages/bitsandbytes/cuda_setup/main.py:149: UserWarning: WARNING: The following directories listed in your path were found to be non-existent: {PosixPath('//172.28.0.1'), PosixPath('http'), PosixPath('8013')}
  warn(msg)
/usr/local/lib/python3.10/dist-packages/bitsandbytes/cuda_setup/main.py:149: UserWarning: WARNING: The following directories listed in your path were found to be non-existent: {PosixPath('//colab.research.google.com/tun/m/cc483011

CUDA_SETUP: WARNING! libcudart.so not found in any environmental path. Searching in backup paths...
CUDA SETUP: CUDA runtime path found: /usr/local/cuda/lib64/libcudart.so
CUDA SETUP: Highest compute capability among GPUs detected: 7.5
CUDA SETUP: Detected CUDA version 122
CUDA SETUP: Loading binary /usr/local/lib/python3.10/dist-packages/bitsandbytes/libbitsandbytes_cuda122.so...


#Setting up the necessary API keys for OpenAI, HuggingFace and LangChain

To get LangChain API key u have to find it from LangSmith website

In [35]:
os.environ["OPENAI_API_KEY"] = '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'
os.environ['HUGGINGFACEHUB_API_TOKEN']="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"
os.environ["LANGCHAIN_TRACING_V2"] = "true"
os.environ["LANGCHAIN_PROJECT"] = f"QnA-ancient-history"
os.environ["LANGCHAIN_ENDPOINT"] = "https://api.smith.langchain.com"
os.environ["LANGCHAIN_API_KEY"] = "ls__02b46e3c0b6e44d19aaa24d65b65edde"

#Loading Data

In [5]:
file_path='/content/chunks_and_questions-ancient_history.json'
data = json.loads(Path(file_path).read_text())

In [6]:
pprint(data[0])

{'chunk_id': 0,
 'chunk_questions': ['Why is the study of ancient Indian history important?',
                     'What does ancient Indian history tell us about the '
                     'earliest cultures in India?',
                     'How did ancient Indians make their life secure and '
                     'settled?',
                     'In what ways did ancient Indians utilize natural '
                     'resources?',
                     'What activities were undertaken by ancient inhabitants '
                     'for their livelihood?',
                     'What were some industries practiced by ancient Indians?',
                     'How did the ancient inhabitants establish villages, '
                     'cities, and kingdoms?',
                     'What is the significance of writing in determining '
                     'civilization?',
                     'Are the Indian scripts and languages used today derived '
                     'from ancient ones?'],

In [7]:
# Define the metadata extraction function.
def metadata_func(record: dict, metadata: dict) -> dict:

    metadata["chunk_id"] = record.get("chunk_id")
    metadata["chunk_questions"] = record.get("chunk_questions")

    return metadata


loader = JSONLoader(
    file_path='/content/chunks_and_questions-ancient_history.json',
    jq_schema='.[]',
    content_key="context_chunk",
    metadata_func=metadata_func
)

data = loader.load()

In [8]:
pprint(data[0])

Document(page_content='1\nThe Significance of\nAncient Indian History\nThe study of ancient Indian history is important for several reasons. It tells us\nhow, when, and where people developed the earliest cultures in India, how they\nbegan undertaking agriculture and stock raising which made life secure and\nsettled. It shows how the ancient Indians discovered and utilized natural\nresources, and how they created the means for their livelihood. We get an idea of\nhow the ancient inhabitants made arrangements for food, shelter, and transport,\nand learn how they took to farming, spinning, weaving, metalworking, and the\nlike, how they cleared forests, founded villages, cities, and eventually large\nkingdoms.\nPeople are not considered civilized unless they know how to write. The\ndifferent forms of writing prevalent in India today are all derived from the\nancient scripts. This is also true of the languages that we speak today', metadata={'source': '/content/chunks_and_questions-ancient

In [9]:
print(data[0].page_content)

1
The Significance of
Ancient Indian History
The study of ancient Indian history is important for several reasons. It tells us
how, when, and where people developed the earliest cultures in India, how they
began undertaking agriculture and stock raising which made life secure and
settled. It shows how the ancient Indians discovered and utilized natural
resources, and how they created the means for their livelihood. We get an idea of
how the ancient inhabitants made arrangements for food, shelter, and transport,
and learn how they took to farming, spinning, weaving, metalworking, and the
like, how they cleared forests, founded villages, cities, and eventually large
kingdoms.
People are not considered civilized unless they know how to write. The
different forms of writing prevalent in India today are all derived from the
ancient scripts. This is also true of the languages that we speak today


In [10]:
max_chunk_size = 0
for i in range(len(data)):
  max_chunk_size = max(len(data[i].page_content), max_chunk_size)
print(max_chunk_size)

1146


#Text Splitting (Not implemented in this scenario)

In [11]:
"""
text_splitter = RecursiveCharacterTextSplitter(
    chunk_size=1146,
    length_function=len,
    add_start_index = True,
)
text_splitter = CharacterTextSplitter(
    separator = "\n",
    chunk_size = 300,
    chunk_overlap  = 100
)
"""

'\ntext_splitter = RecursiveCharacterTextSplitter(\n    chunk_size=1146,\n    length_function=len,\n    add_start_index = True,\n)\ntext_splitter = CharacterTextSplitter(\n    separator = "\n",\n    chunk_size = 300,\n    chunk_overlap  = 100\n)\n'

In [12]:
#docs = text_splitter.split_documents(data)
docs = data

#Chunk Embedding

Was exploring with open source embeddings like **mpnet-base-v2** and **instructor-xl**

In [13]:
"""
model_name = "sentence-transformers/all-mpnet-base-v2"
model_kwargs = {"device": "cuda"}

embeddings = HuggingFaceEmbeddings(model_name=model_name, model_kwargs=model_kwargs)
"""

'\nmodel_name = "sentence-transformers/all-mpnet-base-v2"\nmodel_kwargs = {"device": "cuda"}\n\nembeddings = HuggingFaceEmbeddings(model_name=model_name, model_kwargs=model_kwargs)\n'

In [14]:
from langchain.embeddings import HuggingFaceInstructEmbeddings

model_name="hkunlp/instructor-xl"
model_kwargs = {"device": "cuda"}

embeddings = HuggingFaceInstructEmbeddings(model_name=model_name, model_kwargs=model_kwargs)

/usr/local/lib/python3.10/dist-packages/huggingface_hub/utils/_token.py:88: UserWarning: 
The secret `HF_TOKEN` does not exist in your Colab secrets.
To authenticate with the Hugging Face Hub, create a token in your settings tab (https://huggingface.co/settings/tokens), set it as secret in your Google Colab and restart your session.
You will be able to reuse this secret in all of your notebooks.
Please note that authentication is recommended but still optional to access public models or datasets.
  warnings.warn(


.gitattributes:   0%|          | 0.00/1.48k [00:00<?, ?B/s]

1_Pooling/config.json:   0%|          | 0.00/270 [00:00<?, ?B/s]

2_Dense/config.json:   0%|          | 0.00/116 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/3.15M [00:00<?, ?B/s]

README.md:   0%|          | 0.00/66.3k [00:00<?, ?B/s]

config.json:   0%|          | 0.00/1.52k [00:00<?, ?B/s]

config_sentence_transformers.json:   0%|          | 0.00/122 [00:00<?, ?B/s]

pytorch_model.bin:   0%|          | 0.00/4.96G [00:00<?, ?B/s]

sentence_bert_config.json:   0%|          | 0.00/53.0 [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/2.20k [00:00<?, ?B/s]

spiece.model:   0%|          | 0.00/792k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/2.42M [00:00<?, ?B/s]

tokenizer_config.json:   0%|          | 0.00/2.40k [00:00<?, ?B/s]

modules.json:   0%|          | 0.00/461 [00:00<?, ?B/s]

load INSTRUCTOR_Transformer
max_seq_length  512


#Storing Data in Vectorstores

In [15]:
#db = FAISS.from_documents(docs, embedding=OpenAIEmbeddings())
ids = [str(i) for i in range(1, len(docs) + 1)]
db = FAISS.from_documents(docs, embedding=embeddings, ids=ids)

Checking if our vectorstore is fetching relevant documents

In [16]:
question = "What is the origin of the different forms of writing in India?"
searchDocs = db.similarity_search(question)
print(searchDocs[0].page_content)

The
different forms of writing prevalent in India today are all derived from the
ancient scripts. This is also true of the languages that we speak today. The
languages we use have roots in ancient times, and have developed through the
ages.
Unity in Diversity
Ancient Indian history is interesting because many races and tribes intermingled
in early India. The pre-Aryans, the Indo-Aryans, the Greeks, the Scythians, the
Hunas, the Turks, and others made India their home. Each ethnic group
contributed its mite to the evolution of the Indian social system, art and
architecture, language and literature. All these peoples and their cultural traits
commingled so inextricably that currently they can be clearly identified in their

original form.
A remarkable feature of ancient Indian culture has been the commingling of
cultural elements from the north and south, and from the east and west


#Loading LLM models

Google Flan T5 and Meta LLama 2 7b chat are explored in this scenario

In [17]:
#model_id  = "google/flan-t5-large"
#tokenizer = AutoTokenizer.from_pretrained(model_id)
#model = AutoModelForCausalLM.from_pretrained(model_id, device_map='auto')
#question_answerer = pipeline("text-generation", model=model, tokenizer=tokenizer)

In [18]:
model_id = 'meta-llama/Llama-2-7b-chat-hf'
device = f'cuda:{cuda.current_device()}' if cuda.is_available() else 'cpu'

# set quantization configuration to load large model with less GPU memory
# this requires the `bitsandbytes` library
bnb_config = BitsAndBytesConfig(
    load_in_4bit=True,
    bnb_4bit_quant_type='nf4',
    bnb_4bit_use_double_quant=True,
    bnb_4bit_compute_dtype=bfloat16
)

# begin initializing HF items, need auth token for these
hf_auth = "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"
model_config = AutoConfig.from_pretrained(
    model_id,
    use_auth_token=hf_auth
)

model = LlamaForCausalLM.from_pretrained(
    model_id,
    trust_remote_code=True,
    config=model_config,
    quantization_config=bnb_config,
    device_map='auto',
    use_auth_token=hf_auth
)
model.eval()
print(f"Model loaded on {device}")

config.json:   0%|          | 0.00/614 [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/modeling_utils.py:2193: FutureWarning: The `use_auth_token` argument is deprecated and will be removed in v5 of Transformers.
  warnings.warn(
The argument `trust_remote_code` is to be used with Auto classes. It has no effect here and is ignored.


model.safetensors.index.json:   0%|          | 0.00/26.8k [00:00<?, ?B/s]

model-00001-of-00002.safetensors:   0%|          | 0.00/9.98G [00:00<?, ?B/s]

model-00002-of-00002.safetensors:   0%|          | 0.00/3.50G [00:00<?, ?B/s]

Loading checkpoint shards:   0%|          | 0/2 [00:00<?, ?it/s]

generation_config.json:   0%|          | 0.00/188 [00:00<?, ?B/s]

Model loaded on cuda:0


In [19]:
tokenizer = AutoTokenizer.from_pretrained(
    model_id,
    use_auth_token=hf_auth
)

tokenizer_config.json:   0%|          | 0.00/1.62k [00:00<?, ?B/s]

/usr/local/lib/python3.10/dist-packages/transformers/tokenization_utils_base.py:1714: FutureWarning: The `use_auth_token` argument is deprecated and will be removed in v5 of Transformers.
  warnings.warn(


tokenizer.model:   0%|          | 0.00/500k [00:00<?, ?B/s]

tokenizer.json:   0%|          | 0.00/1.84M [00:00<?, ?B/s]

special_tokens_map.json:   0%|          | 0.00/414 [00:00<?, ?B/s]

**Pipline for generating answers**

In [20]:
question_answerer = pipeline(
    model=model, tokenizer=tokenizer,
    return_full_text=True,  # langchain expects the full text
    task='text-generation',
    # we pass model parameters here too
    temperature=0.0,  # 'randomness' of outputs, 0.0 is the min and 1.0 the max
    max_new_tokens=512,  # mex number of tokens to generate in the output
    repetition_penalty=1.1,  # without this output begins repeating
)

**Pipline for evaluating generated answers**

In [21]:
evaluater = pipeline(
    model=model, tokenizer=tokenizer,
    return_full_text=True,  # langchain expects the full text
    task='text-generation',
    # we pass model parameters here too
    temperature=0.0,  # 'randomness' of outputs, 0.0 is the min and 1.0 the max
    max_new_tokens=512,  # mex number of tokens to generate in the output
    repetition_penalty=1.1,  # without this output begins repeating
)

In [22]:
local_llm = HuggingFacePipeline(pipeline=question_answerer)

In [23]:
eval_llm = HuggingFacePipeline(pipeline=evaluater)

#Retriever Definition

In [24]:
retriever = db.as_retriever()

In [25]:
docs = retriever.get_relevant_documents("What connection do the people living on the Sindhu or the Indus have with the naming of the country?")
print(docs[0].page_content)

Our country shows a deep
underlying unity despite great diversity.
The ancients strove for unity. The Indian subcontinent was geographically
well defined and its geographical unity was supplemented by cultural
integration. Though there existed many states, languages, cultures, and
communities, gradually people developed territorial identity. The states or
territorial units, called 
janapada
s, were named after different tribes. However,
the country as a whole came to be named Aryavarta after the dominant cultural
community called the Aryans. Aryavarta denoted northern and central India and
extended from the eastern to the western sea coasts. The other name by which
India was better known was Bharatavarsha or the land of the Bharatas


In [26]:
retriever = db.as_retriever(search_kwargs={"search_type":"mmr","k": 3})

#Prompt Definition

In [27]:
def format_docs(docs):
    return "\n\n".join(doc.page_content for doc in docs)

template = """Assume you are a history teacher. Use the following pieces of context
   to answer the question at the end. If you don't know the answer, just say that
   you don't know, don't try to make up an answer.Use three sentences maximum and
   keep the answer as concise as possible.

{context}

Question: {question}

Helpful Answer:"""

custom_rag_prompt = PromptTemplate(
    input_variables=["question"],
    template=template
)

prompt = hub.pull("rlm/rag-prompt")

#Chain Definition



In [28]:
from langchain.chains.question_answering import load_qa_chain
rag_chain = (
    {"context": retriever | format_docs, "question": RunnablePassthrough()}
    | prompt
    | local_llm
    | StrOutputParser()
)
custom_rag_chain = (
    {"context": retriever | format_docs, "question": RunnablePassthrough()}
    | custom_rag_prompt
    | local_llm
    | StrOutputParser()
)
qa = RetrievalQA.from_chain_type(llm=local_llm, chain_type="stuff", retriever=retriever, return_source_documents=True)

#Sample User Question Input

In [29]:
question = "What connection do the people living on the Sindhu or the Indus have with the naming of the country?"

#Answers from various Chains

In [30]:
question = question
result = qa.invoke({"query": question})

In [31]:
pprint(result['result'])
print('*'*100)
for i in range(len(result['source_documents'])):
  content = result['source_documents'][i].page_content
  print(f'Document {i} : {content}')

(' The people living on the Sindhu or the Indus are connected to the naming of '
 'the country because the term "Hind" or "Hindu" is derived from the Sanskrit '
 'term "Sindhu," which refers to the Indus River. This term was used by '
 'foreigners to refer to the entire country, and eventually, the name "India" '
 'was adopted.')
****************************************************************************************************
Document 0 : Our country shows a deep
underlying unity despite great diversity.
The ancients strove for unity. The Indian subcontinent was geographically
well defined and its geographical unity was supplemented by cultural
integration. Though there existed many states, languages, cultures, and
communities, gradually people developed territorial identity. The states or
territorial units, called 
janapada
s, were named after different tribes. However,
the country as a whole came to be named Aryavarta after the dominant cultural
community called the Aryans. Aryava

In [32]:
pprint(custom_rag_chain.invoke(question))

(' The people living on the Sindhu or the Indus River were the first to come '
 'into contact with foreigners, leading to the naming of the country as '
 '"India" by these foreigners.')


In [33]:
pprint(rag_chain.invoke(question))

(' The people living on the Sindhu or the Indus had a significant connection '
 'with the naming of the country, as the term "Hindu" is derived from the '
 'Sanskrit term "Sindhu," which refers to the Indus River. The term "Hindu" '
 'was initially used to refer to a specific territorial unit on the Indus '
 'River, and later came to be used as a broader term to refer to the entire '
 'country.')


#Answer Evaluation using Question - Answer pair

Loaded the ground truth dataset in LangSmith


In [36]:
warnings.filterwarnings("ignore")

# Define the evaluators to apply
eval_config = smith.RunEvalConfig(
    evaluators=[
        "cot_qa",
        "qa",
        "context_qa",
    ],
    prediction_key="result",
    eval_llm = eval_llm
)
load_dotenv()
client = langsmith.Client()
chain_results = client.run_on_dataset(
    dataset_name="question_and_answers-ancient-history",
    llm_or_chain_factory=qa,
    evaluation=eval_config,
    project_name="faiss-mmr-retrievalqa-llama2",
    verbose=True,
)

View the evaluation results for project 'faiss-mmr-retrievalqa-llama2' at:
https://smith.langchain.com/o/7514939f-03f4-498d-ae64-0b4270db2e96/datasets/5807e033-1e13-4412-8411-9ab32c4f7df8/compare?selectedSessions=0f1c6b15-43b6-4a7a-a990-2f4ac7a8d4b5

View all tests for Dataset question_and_answers-ancient-history at:
https://smith.langchain.com/o/7514939f-03f4-498d-ae64-0b4270db2e96/datasets/5807e033-1e13-4412-8411-9ab32c4f7df8
[--------------------------------------->          ] 95/120

ERROR:langchain_core.tracers.evaluation:Error evaluating run 186d70a4-c214-4fb2-9783-feebb2c75575 with StringRunEvaluatorChain: LangSmithError('Failed to POST https://api.smith.langchain.com/feedback in LangSmith API. HTTPError(\'400 Client Error: Bad Request for url: https://api.smith.langchain.com/feedback\', \'{"detail":"Feedback score is required for continuous feedback"}\')')
Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/langsmith/utils.py", line 88, in raise_for_status_with_text
    response.raise_for_status()
  File "/usr/local/lib/python3.10/dist-packages/requests/models.py", line 1021, in raise_for_status
    raise HTTPError(http_error_msg, response=self)
requests.exceptions.HTTPError: 400 Client Error: Bad Request for url: https://api.smith.langchain.com/feedback

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/usr/local/lib/python3.10/dist-packages/langsmith/client.py", lin

[------------------------------------------------->] 120/120
 Experiment Results:
        feedback.COT Contextual Accuracy  feedback.correctness  feedback.Contextual Accuracy error  execution_time                                run_id
count                             119.00                119.00                        119.00     0          120.00                                   120
unique                               NaN                   NaN                           NaN     0             NaN                                   120
top                                  NaN                   NaN                           NaN   NaN             NaN  f7000c6f-6b8d-41ce-996a-92e1b04c7ddb
freq                                 NaN                   NaN                           NaN   NaN             NaN                                     1
mean                                0.44                  0.66                          0.55   NaN           74.47                                   NaN


#Delete vectorstore

In [37]:
db.delete(ids=ids)

True

##The **Correctness Accuracy achieved from this implementation is 66%**, Contextual Accuracy 55% and Contextual Accuracy wuth reasioning 44%.